### PR TITLE
Issue #24: Add braces to if statement

### PR DESCRIPTION
**Pull request by HEYGUL**
#24 implementation

### CODE REVIEW

**Comment by HEYGUL** [comment]
I open the PR as soon as possible to take opportunity to get feedback as soon as possible during the implement of my very first `Abracadabra` feature.
Feel free to comment and/or give advice. 👍

File: src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    return;
+  }
+
+  await editor.write(updatedCode.code);
+}
+
+function hasIfStatementToAddBraces(ast: t.AST, selection: Selection): boolean {
+  let result = false;
+
+  t.traverseAST(ast, {
```
Comment (nicoespeon) [inline src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.ts]: Here we can use the `createVisitor()` method to get rid of the duplicated logic.

We would replace this with 
```js
t.traverseAST(ast, createVisitor(selection, () => (result = true)));
```

File: src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.test.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  testEach<{ code: Code; selection?: Selection; expected: Code }>(
+    "should add braces to if statement",
+    [
+      {
+        description: "basic scenario",
+        code: "if (!isValid) return;",
+        selection: Selection.cursorAt(0, 17),
+        expected: `if (!isValid) {
+  return;
+}`
```
Comment (nicoespeon) [inline src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.test.ts]: 👍 That's great!

Now can we think about other use-cases? What would happen for:

```js
if (isValid) 
  doSomething();
else
  doAnotherThing();
```

Ideally, it should only add braces to the `if` statement (= consequent), so I think it already works. But then, if we trigger that from the `else` (= alternate), I'd expect to add braces on the `else` statement instead.

That might be more tricky to implement. Let me know if you need help to figure this out. 

And this one should not change (which, I think, is already the case):

```js
if (isValid) {
  doSomething();
}
```

File: src/refactorings/add-braces-to-if-statement/index.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    operation: addBracesToIfStatement,
+    title: "Add Braces To If Statement"
+  },
+  actionProvider: {
+    message: "Add braces to if statement",
+    canPerform: hasIfStatementToAddBraces
+  }
+};
+
+export default config;
```
Comment (nicoespeon) [inline src/refactorings/add-braces-to-if-statement/index.ts]: All of this is very nice. The essentials of the feature seems to be here. We need to declare the refactoring in the `extension.ts` file and in the `package.json`, like the others, so it becomes available. 

You should be able to see it if you run the extension host debugger (`F5`).

File: src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.test.ts (hunk 2 above)
Comment (HEYGUL) [reply src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.test.ts]: Yes if have a full list of use-cases. My first goal was to get a first understanding of AST manipulation and get my first green test (TDD forever).

File: src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.ts (hunk 1 above)
Comment (HEYGUL) [reply src/refactorings/add-braces-to-if-statement/add-braces-to-if-statement.ts]: I agree.

File: src/refactorings/add-braces-to-if-statement/index.ts (hunk 3 above)
Comment (HEYGUL) [reply src/refactorings/add-braces-to-if-statement/index.ts]: Thank you, I will have a look to those 2 files.

**Comment by nicoespeon** [comment]
@HEYGUL do you need my help? No pressure, but I want to check in case you'd be blocked with something—that's fine if you just didn't have time to move on this, there's no hurry.

**Comment by HEYGUL** [comment]
> @HEYGUL do you need my help? No pressure, but I want to check in case you'd be blocked with something—that's fine if you just didn't have time to move on this, there's no hurry.

Hey @nicoespeon, I just did not have time till yesterday. I took into account some of your feedbacks and I am working on the `else` selection scenario :-)

**Comment by nicoespeon** [comment]
@HEYGUL perfect, let me know when you're done 👍

**Comment by HEYGUL** [comment]
I implement both consequent and alternate selections to add braces to, but if you select the `else` or `if` keywords themselves, the `selection.isInsideNode` function returns false.
I am not sure how this refactoring shall work 🤔 @nicoespeon I need your advice and point of view about.

I updated README for this refactoring but I do not know how you generate the demo gifs, any tips are welcome.

**Comment by nicoespeon** [comment]
@HEYGUL I'll have a look at this next week when I'm back to work. Thanks for contributing 😉 

As for the gifs, I can create them. But if you want to know, I'm using https://recordit.co/ and it's super convenient!

**Comment by nicoespeon** [comment]
@HEYGUL your implementation was very good. It didn't miss a lot to complete the feature.

Indeed, I'd expect the UX to handle when the cursor is on the `if` or `else` keyword, so usage is more intuitive. It turns out you weren't far from solving this. In fact, instead of checking for `isInsideNode()`, we had to check the position of the start of the cursor:

- if it's before the end of the `consequent` node, then it means we're on the `if` part
- else (it's after the end of the `consequent` node), then it means we're on the `else` part

So the code becomes:

```js
      if (!t.isSelectableNode(path.node.consequent)) return;
      const endOfConsequent = Position.fromAST(path.node.consequent.loc.end);

      if (selection.start.isBefore(endOfConsequent)) {
        path.node.consequent = statementWithBraces(path.node.consequent);
        return;
      }

      if (path.node.alternate) {
        path.node.alternate = statementWithBraces(path.node.alternate);
      }
```

And the `statementWithBraces()`:

```js
function statementWithBraces(node: t.Statement): t.Statement {
  return t.isBlockStatement(node) ? node : t.blockStatement([node]);
}
```

Finally, I checked for a nasty scenario: **nested if statements**. 
With current implementation, it would always take the outer if statement and stop.

To solve this, we need to check if a children if statement would match. That's a classic check we need to do, so I think there's a pattern I could refactor. But I didn't yet. Nevermind, I implemented the fix along with tests.

IMO, we're good to go. I'll update the Changelog & README. This will probably be released by the end of the week.

Do you have any question before I merge?

File: package.json [hunk 4]
```diff
@@ -632,6 +638,10 @@
         {
           "command": "abracadabra.splitIfStatement",
           "when": "editorLangId == typescriptreact"
+        },
+        {
+          "command": "abracadabra.addBracesToIfStatement",
+          "when": "editorLangId == javascript"
```
Comment (nicoespeon) [inline package.json]: Can you duplicate this command declaration to enable it for all the other 3 languages?

If you look closely, you'll notice that each refactoring is declared for the 4 languages (js, jsx, ts, tsx). That's a lot of boilerplate, but I didn't put an abstraction on top to simplify that—yet.

Comment (nicoespeon) [reply package.json]: Nevermind, it was a simple boilerplate: I did it so I can merge your PR and release this 😉

**Comment by nicoespeon** [comment]
Merging this. Thanks for your contribution @HEYGUL 👍

**Pull request merged**